### ISSUE TITLE
“pull: no available source device” on Xubuntu, even though the file exists on Android

### ISSUE THREAD

**Issue by marcelpaulo** [syncthing/syncthing-android#351]
Issue [#2406](https://forum.syncthing.net/t/pull-no-available-source-device-on-xubuntu-even-though-the-file-exists-on-an-android/2406) discussed on discourse before posting here.

The environment:

- an android phone (KitKat 4.4.4), `thelonious`, with shared folder `camera`:
![screenshot_2015-04-11-11-44-03](https://cloud.githubusercontent.com/assets/10074769/7101848/2f863340-e042-11e4-9ada-5ae92fd6c6e3.png)
- a xubuntu (14.10) desktop, `monk`, sharing `camera`:
![syncthing gtk_002](https://cloud.githubusercontent.com/assets/10074769/7101851/6ec97936-e042-11e4-85ae-5ed2eafe5f32.jpg)
- syncthing-android 0.5.28
- syncthing daemon 0.30.0
- syncthing-gtk 0.6.4

Steps to reproduce the error:

1. Take a photo with the phone
2. Crop photo using `QuickPic`
3. The original photo is synced to the desktop but the cropped photo gets stuck

Daemon messages on the desktop:
```
[2L7Y5] 11:51:25 INFO: Puller (folder "camera", file "IMG_20150411_115059414~01.jpg"): pull: no available source device
[2L7Y5] 11:51:25 WARNING: Folder "camera" isn't making progress - check logs for possible root cause. Pausing puller for 1m0s.
```

Files on the phone:
![screenshot_2015-04-11-11-51-37](https://cloud.githubusercontent.com/assets/10074769/7101878/83158f5a-e043-11e4-83c6-b16417b5737e.png)

Files on the desktop:
![mc paulo monk -thelonious-camera_003](https://cloud.githubusercontent.com/assets/10074769/7101880/a73bffa4-e043-11e4-8491-6ba0524a997b.jpg)

Shared folder on the desktop:
![syncthing gtk_005](https://cloud.githubusercontent.com/assets/10074769/7101912/b43a8b5c-e044-11e4-9496-42e7b7167866.jpg)

Shared folder on the phone:
![screenshot_2015-04-11-12-17-51](https://cloud.githubusercontent.com/assets/10074769/7101918/e20ae752-e044-11e4-8691-e8f1eade402b.png)

The only way to resync the folder:

1. Delete both files on the phone
2. Rescan the folder on the phone

I tried to get a logcat following [your instructions](https://github.com/syncthing/syncthing-android/blob/master/CONTRIBUTING.md), but my phone isn't rooted, so none worked.

How can I help debug this further ?

***EDIT:*** Ran `aLogcat` after reproducing the error, but very few lines from `/dev/log/main` and `/dev/log/system` were shown, none related to syncthing

**Comment by Nutomic** [comment]
This doesn't seem like an Android-specific problem to me, you should probably report this upstream.

For the logs, have you tried to filter for "SyncthingNativeCode"? Also, in the forum you said that you put "STTRACE=xyz" into debug options, but that only needs the "xyz" part (you should drop the "STTRACE=".

**Comment by marcelpaulo** [comment]
My cell phone runs on Android 4.4.4 and is not rooted, so aLogcat didn't
show anything related to syncthing.

Although the error doesn't seem specific to Android, it's easily reproduced
there: it always happens when I crop a photo. Not sure how we could force
the error on a desktop. The fact is: keeping an Android photo folder synced
is not viable with this error.

**Comment by Nutomic** [comment]
Can you use `adb logcat` from a pc for the logging?

http://wiki.cyanogenmod.org/w/Doc:_debugging_with_logcat

**Comment by marcelpaulo** [comment]
A confession is in order here: when I read  your Wiki, I thought adb ran on
the phone. My ignorance! I'll give it a try and post back the results.

Thanks for the link!

**Comment by Nutomic** [comment]
Any suggestions to improve the text? PR welcome :)

**Comment by marcelpaulo** [comment]
Oh, no, it's clear and well written. I am to blame: my Android experience
is next to nothing.

**Comment by Nutomic** [comment]
Imo that's a perfectly valid reason to improve it. In the future, #311 should make this easier.

**Comment by marcelpaulo** [comment]
I feel ashamed for not having done this before, it was easier than I thought. For the record, to use `adb` on Xubuntu 14.10, it was just a question of (besides enabling `Developer options` on the phone):

1. Installing package `android-tools-adb`
2. Connecting the phone to the desktop via USB cable
3. Run `adb devices`
4. Accept the desktop fingerprint on the phone

Here's what I did to reproduce the problem and capture the logs:

1. *[phone]* Write `scanner` on `Debug Options`. Recommended by @Nutomic and Audrius Butkevicius [here](https://forum.syncthing.net/t/pull-no-available-source-device-on-xubuntu-even-though-the-file-exists-on-an-android/2406/13?u=marcelpaulo)
2. *[desktop]* Start syncthing: `STTRACE=model syncthing &> >(tee desktop.log)`. Recommended by Audrius Butkevicius [here](https://forum.syncthing.net/t/pull-no-available-source-device-on-xubuntu-even-though-the-file-exists-on-an-android/2406/11?u=marcelpaulo)
3. *[desktop]* Start logcat: `adb logcat | grep -i syncthing &> >(tee phone.log)`
4. *[phone]* Take a photo
5. *[phone]* Crop the photo with QuickPic
6. *[desktop]* Confirm error, then stop syncthing and logcat

Files on the phone:
![screenshot_2015-04-14-06-39-52](https://cloud.githubusercontent.com/assets/10074769/7134781/8aabdfbe-e276-11e4-87be-ee64b3dd4fde.png)

Files on the desktop:
![mc paulo monk -thelonious-camera_001](https://cloud.githubusercontent.com/assets/10074769/7134788/9ac6ad66-e276-11e4-8931-9f6813c597fd.jpg)

The logs and config:

[desktop.log](https://docs.google.com/file/d/0B6eQjhXV-AquRVFydFhKQlU4N0E/edit?usp=drivesdk)
[phone.log](https://docs.google.com/file/d/0B6eQjhXV-AqubjFubElYTDNjbnM/edit?usp=drivesdk)
[desktop config.xml](https://docs.google.com/file/d/0B6eQjhXV-Aqubk9XbzVLams0Rzg/edit?usp=drivesdk)
[phone config.xml](https://docs.google.com/file/d/0B6eQjhXV-AquY1NrMVBheFJwOVU/edit?usp=drivesdk)

I hope I haven't missed anything.

Thanks for your patience and helpful tips, @Nutomic ! Let me know if there's anything else I can help with to debug this.

**Comment by calmh** [comment]
I can see this happening if the contents of the file changes, but not the modification time or the size. Then we have no reason to rehash the file, so will announce the old hashes. The fetching side will then come up with this rather unintuitive error.

**Comment by marcelpaulo** [comment]
Fixed in latest 0.11 branch. Can't reproduce the issue anymore.

Using:

syncthing: 0.11.0-rc1
syncthing-gtk: 0.692
syncthing-android: v0.11.0-beta6+5-gba4a6fc, Linux (ARM)
